Fix bugs in line vert shader

JakubValtar opened the pull request:
Fixes #5181
Fixes #5182
Fixes #5183
Fixes #5194

point vert updated with the code from line vert to keep them in sync

benfry commented:
@codeanticode want to review before we merge?

benfry commented:
Oh look, Github now has a reviewer thing… Don't mind me, I'll just be returning to my hole called 2013.

codeanticode commented:
ha, no worries, I will take a look now!

codeanticode commented:
This code is excellent, much cleaner with a nice derivation of the tangent vector that factors out the problematic division by w, and without apparent edge cases. I will run a few tests and will report back, but don't see any side effects so far.

codeanticode commented on core/src/processing/opengl/shaders/LineVert.glsl:
```diff
@@ -34,54 +34,64 @@ attribute vec4 color;
+  // convert from screen to clip (derived from clip to screen above)
+  vec2 noPerspScale = p.w / (0.5 * viewport.zw);
+
+  gl_Position.xy = p.xy + offset.xy * mix(noPerspScale, perspScale, perspective > 0);
```
@JakubValtar the glsl compiler on Mac does not automatically convert perspective > 0 into a number, so it gives an error bc does not find a mix(float, float, bool) function. This works:

mix(noPerspScale, perspScale, float(perspective > 0))

JakubValtar replied:
Ah, the overload is in GLSL docs, but looks like only for GLSL 4.50. Try to leave there only `perspective` without anything, it is integer of 0 or 1 and should work on it's own.

JakubValtar replied:
Or `float(perspective)` if the int does not work either.

JakubValtar commented:
@codeanticode I modified the shaders and checked that they run with `#version 110` which means that they should run on all cards supporting OpenGL 2.0 or newer. I also modified the version check in shader preprocessor to ignore comments (something I've ran into while testing and which we should have in place).

codeanticode commented:
@JakubValtar thanks for the last tweaks

@benfry this PR is ready to merge

JakubValtar commented:
@codeanticode Thanks for the review!

JakubValtar commented:
Found another one which I accidentally fixed.

Fixes #5194

benfry commented:
Thanks gentlemen…

JakubValtar commented:
*tips hat*

On Tue, 22 Aug 2017, 15:48 Ben Fry <notifications@github.com> wrote:

> Thanks gentlemen…
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/processing/processing/pull/5184#issuecomment-324031601>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADB6ikzOt1ZUs98U4DeSqDd3dfDBFm8cks5satwegaJpZM4OeQtI>
> .
>
